nvflare/apis/dxo_filter.py

```diff
@@ -0,0 +1,136 @@
+        """
+        Filter.__init__(self)
+
+        if supported_data_kinds and not isinstance(supported_data_kinds, list):
+            raise ValueError(f"supported_data_kinds must be a list of str but got {type(supported_data_kinds)}")
+
+        if data_kinds_to_filter and not isinstance(data_kinds_to_filter, list):
+            raise ValueError(f"data_kinds_to_filter must be a list of str but got {type(data_kinds_to_filter)}")
+
+        if supported_data_kinds and data_kinds_to_filter:
+            if not all(dk in supported_data_kinds for dk in data_kinds_to_filter):
+                raise ValueError(f"invalid data kinds: {data_kinds_to_filter}. Only support {data_kinds_to_filter}")
+
+        if not data_kinds_to_filter:
+            data_kinds_to_filter = supported_data_kinds
+
+        self.data_kinds = data_kinds_to_filter
+
+    def process(self, shareable: Shareable, fl_ctx: FLContext):
+        rc = shareable.get_return_code()
+        if rc != ReturnCode.OK:
+            # don't process if RC not OK
+            return shareable
+
+        try:
+            dxo = from_shareable(shareable)
+        except:
+            # not a DXO based shareable - pass
+            return shareable
+
+        if dxo.data is None:
+            self.log_debug(fl_ctx, "DXO has no data to filter")
+            return shareable
+
+        start = [dxo]
+        self._filter_dxos(start, shareable, fl_ctx)
+        result_dxo = start[0]
+        return result_dxo.update_shareable(shareable)
+
+    @abstractmethod
+    def process_dxo(self, dxo: DXO, shareable: Shareable, fl_ctx: FLContext) -> Union[None, DXO]:
+        """Subclass must implement this method to filter the provided DXO
+
+        Args:
+            dxo: the DXO to be filtered
+            shareable: the shareable that the dxo belongs to
+            fl_ctx: the FL context
+
+        Returns:
+            A DXO object that is the result of the filtering, if filtered;
+            None if not filtered.
+
+        """
+        pass
+
+    def _apply_filter(self, dxo: DXO, shareable, fl_ctx: FLContext) -> DXO:
+        if not dxo.data:
+            self.log_debug(fl_ctx, "DXO has no data to filter")
+            return dxo
+
+        filter_name = self.__class__.__name__
+        result = self.process_dxo(dxo, shareable, fl_ctx)
+        if not result:
+            # not filtered
+            result = dxo
+        elif not isinstance(result, DXO):
+            raise RuntimeError(f"Result from {filter_name} is {type(result)} - must be DXO")
+        else:
+            if result != dxo:
+                # result is a new DXO - copy filter history from original dxo
+                result.add_filter_history(dxo.get_filter_history())
+            result.add_filter_history(filter_name)
+
+            chain_type = self.get_prop(FilterContextKey.CHAIN_TYPE, "?")
+            source = self.get_prop(FilterContextKey.SOURCE, "?")
```

Comment:
why don't we just make these two be instance variables for the `Filter` class?

Reply:
either way. it's subjective.